### ISSUE TITLE
Incorrect type inference in 'match' expression

### ISSUE THREAD

**Issue by developedby** [HigherOrderCO/Bend#727]
### Reproducing the behavior

Considering the following program
```py
type Box:
  Box {x: Unit}

type Unit:
  Unit

Erase: Box -> None
(Erase (Box/Box x)) = (Erase x)

main: None
main = (Erase (Box/Box Unit/Unit))
```

This should generate a type error in the `Erase` function, since it expects a `Box` value but we're giving it a `Unit`. 

Based on the match expression, the type of `Erase` should be `Box -> var`, while according to the call it should be `Unit -> var`. When unifying these two we expect a type error, which never happens.

The final infered type is `Box -> var`, which specializes to `Box -> None` when checking against the annotation.

### System Settings

Bend commit: 4d6f461d54e43c4aa9bfde14fdc611e235184129

**Comment by developedby** [comment]
The issue was that the type constructors inside of constructor fields were not being resolved and were kept as variables. So, even though the type of the `Box/Box` constructor was being updated to `Unit -> Box` with Unit as a constructor, the type of `x` was still `Unit` as a type variable.

This means that all fields inside of match expressions were being inferred variable types for the nullary constructors in their types.

**Comment by developedby** [comment]
Fixed by #728
